File: MoonEngine-Root/Assets/src/main/java/com/greentree/commons/assets/serializator/manager/AsyncAssetManager.java
package com.greentree.commons.assets.serializator.manager;

import com.greentree.commons.assets.key.AssetKey;
import com.greentree.commons.assets.key.ResourceAssetKey;
import com.greentree.commons.assets.key.ResultAssetKey;
import com.greentree.commons.assets.serializator.request.KeyLoadRequest;
import com.greentree.commons.assets.serializator.request.builder.KeyRequestBuilder;
import com.greentree.commons.assets.serializator.request.builder.KeyRequestBuilderImpl;
import com.greentree.commons.assets.value.Value;
import com.greentree.commons.util.classes.info.TypeInfo;
import com.greentree.commons.util.classes.info.TypeInfoBuilder;

public interface AsyncAssetManager extends ValueAssetManager {
	
	default <T> Value<T> loadAsync(KeyLoadRequest<T> request) {
		return loadAsync(request.loadType(), request.key(), request.getDefault());
	}
	
	default <T> KeyRequestBuilder<T, Value<T>> loadAsync(TypeInfo<T> type) {
		return new KeyRequestBuilderImpl<>(r->loadAsync(r), type);
	}
	
	default <T> KeyRequestBuilder<T, Value<T>> loadAsync(Class<T> cls) {
		final var type = TypeInfoBuilder.getTypeInfo(cls);
		return loadAsync(type);
	}
	
	default <T> Value<T> loadAsync(Class<T> cls, AssetKey key) {
		final var type = TypeInfoBuilder.getTypeInfo(cls);
		return loadAsync(type, key, loadDefault(type));
	}
	
	default <T> Value<T> loadAsync(Class<T> cls, AssetKey key, T def) {
		final var type = TypeInfoBuilder.getTypeInfo(cls);
		return loadAsync(type, key, loadDefault(type, def));
	}
	
	default <T> Value<T> loadAsync(TypeInfo<T> type, AssetKey key) {
		return loadAsync(type, key, loadDefault(type));
	}
	
	<T> Value<T> loadAsync(TypeInfo<T> type, AssetKey key, T def);
	
	default <T> Value<T> loadAsync(Class<T> cls, String path) {
		return loadAsync(cls, new ResourceAssetKey(path), loadDefault(cls));
	}
	
	default <T> Value<T> loadAsync(Class<T> cls, String path, T def) {
		final var type = TypeInfoBuilder.getTypeInfo(cls);
		return loadAsync(type, new ResourceAssetKey(path), loadDefault(type, def));
	}
	
	default <T> Value<T> loadAsync(TypeInfo<T> type, String path) {
		return loadAsync(type, new ResourceAssetKey(path), loadDefault(type));
	}
	
	default <T> Value<T> loadAsync(TypeInfo<T> type, String path, T def) {
		return loadAsync(type, new ResourceAssetKey(path), def);
	}
	
	default <T> Value<T> loadAsync(Class<T> cls, Object obj) {
		return loadAsync(cls, new ResultAssetKey(obj), loadDefault(cls));
	}
	
	default <T> Value<T> loadAsync(Class<T> cls, Object obj, T def) {
		final var type = TypeInfoBuilder.getTypeInfo(cls);
		return loadAsync(type, new ResultAssetKey(obj), loadDefault(type, def));
	}
	
	default <T> Value<T> loadAsync(TypeInfo<T> type, Object obj) {
		return loadAsync(type, new ResultAssetKey(obj), loadDefault(type));
	}
	
	default <T> Value<T> loadAsync(TypeInfo<T> type, Object obj, T def) {
		return loadAsync(type, new ResultAssetKey(obj), def);
	}
	
}

File: MoonEngine/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/manager/AsyncAssetManager.kt
package com.greentree.engine.moon.assets.serializator.manager

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.commons.reflection.info.TypeInfoBuilder
import com.greentree.engine.moon.assets.key.AssetKey
import com.greentree.engine.moon.assets.key.ResourceAssetKey
import com.greentree.engine.moon.assets.key.ResultAssetKey
import com.greentree.engine.moon.assets.provider.AssetProvider
import com.greentree.engine.moon.assets.provider.ConstAssetProvider
import com.greentree.engine.moon.assets.provider.ReduceAssetProvider
import com.greentree.engine.moon.assets.serializator.manager.chain.Chain
import com.greentree.engine.moon.assets.serializator.manager.chain.ChainHandler
import java.lang.Thread.*
import java.util.concurrent.Executors

private var ID = 0
val EXECUTOR = Executors.newFixedThreadPool(
	Runtime.getRuntime().availableProcessors()
) {
	val thread = Thread(it, "async-asset-${ID++}")
	thread.isDaemon = true
	thread.priority = MAX_PRIORITY
	thread
}

data object AsyncHandler : ChainHandler {

	override fun <T : Any> load(chain: Chain, type: TypeInfo<T>, key: AssetKey): AssetProvider<T> {
		val res = chain.loadCache(type, key)
		if(res != null)
			return res
		val default = chain.loadDefault(type, key.type()) ?: return chain.load(type, key)
		val result = ReduceAssetProvider(ConstAssetProvider(default))
		EXECUTOR.submit {
			try {
				result.origin = chain.load(type, key)
			} catch(e: Exception) {
				e.printStackTrace()
			}
		}
		return result
	}
}

fun <T : Any> ChainAssetManager.loadAsyncProvider(type: TypeInfo<T>, key: AssetKey) =
	build(AsyncHandler).load(type, key)

fun <T : Any> ChainAssetManager.loadAsync(type: TypeInfo<T>, key: AssetKey) = wrap(build(AsyncHandler).load(type, key))

fun <T : Any> ChainAssetManager.loadAsync(type: TypeInfo<T>, key: Any) = loadAsync(type, ResultAssetKey(key))
fun <T : Any> ChainAssetManager.loadAsync(type: TypeInfo<T>, resource: String) =
	loadAsync(type, ResourceAssetKey(resource))

fun <T : Any> ChainAssetManager.loadAsync(cls: Class<T>, key: AssetKey) =
	loadAsync(TypeInfoBuilder.getTypeInfo(cls), key)

fun <T : Any> ChainAssetManager.loadAsync(cls: Class<T>, key: Any) = loadAsync(TypeInfoBuilder.getTypeInfo(cls), key)
fun <T : Any> ChainAssetManager.loadAsync(cls: Class<T>, resource: String) =
	loadAsync(TypeInfoBuilder.getTypeInfo(cls), resource)

inline fun <reified T : Any> ChainAssetManager.loadAsync(key: AssetKey) =
	loadAsync(TypeInfoBuilder.getTypeInfo(T::class.java), key)

inline fun <reified T : Any> ChainAssetManager.loadAsync(key: Any) =
	loadAsync(TypeInfoBuilder.getTypeInfo(T::class.java), key)

inline fun <reified T : Any> ChainAssetManager.loadAsync(resource: String) =
	loadAsync(TypeInfoBuilder.getTypeInfo(T::class.java), resource)
